Slow cyberstorm api endpoints (TS-2516)

Roffenlund opened the pull request:
Implement a celery task and caching mechanism for rendering markdown to HTML. This removes the CPU overhead caused by rendering the HTML on every request.

* Add new celery queue
* Add new redis cache
* Implement celery task for handling the markdown render
* Refactor the API views for readme and changelog to use the celery task

MythicManiac commented on django/thunderstore/repository/tasks/markdown.py:
```diff
@@ -0,0 +1,35 @@
+    if cached_html is not None:
+        return cached_html
+
+    with cache.lock(lock_key, timeout=CACHE_LOCK_TIMEOUT, blocking_timeout=None):
```
Locking in the celery task itself isn't really necessary, the task queue has a controlled level of parrallelism and double-writes aren't a problem, which basically means this is more efficient to do without locking (and parallel runs of the same rendering won't cause problems in case it does happen)

MythicManiac commented on django/thunderstore/repository/tasks/markdown.py:
```diff
@@ -0,0 +1,35 @@
+            cache.delete(status_key)
+            raise error
+
+        cache.set(cache_key, html, timeout=DEFAULT_CACHE_EXPIRY)
```
This cache can be permanent, no need to expire immutable data.

That said, if the markdown renderer code changes we _could_ bust the cache if we have a reasonable mechanism for detecting that somehow. I'd not consider it a huge issue even if we don't bust the cache in that case though, as long as we document it somewhere.

Roffenlund committed:
Add celery queue and redis storage for markdown

Add a new queue intended for running tasks that are related to rendering
markdown to html.

Add a redis cache for caching the html once it is rendered so we do not
need to re-run the rendering task but can fetch it from redis.

Refs. TS-2516

Roffenlund committed:
Implement markdown celery task

Implement a celery task for rendering the markdown to html.

Refs. TS-2516

Roffenlund committed:
Implement service for rendering markdown

Implement a service which implements the caching logic around calling
the celery task with the correct variables. The intention of the service
is to be re-usable with views and make unit testing easier.

Refs. TS-2516

Roffenlund committed:
Refactor markdown views

USe the markdown service which is responsible for calling the markdown
task and handling the cache so we do not need to run the markdown logic
in each request, but can fetch the results from cache instead.

Refs. TS-2516

Roffenlund committed:
Rework markdown service and celery task

Rewrite the polling logic in the service to use the get() function from
celery instead to wait for the result of a task. Raise errors when
expected, the .get() function raises either a TimeoutError or Exception.

No need to pass the statys_key to the celery task anymore due to the
changes made in the service.

Update tests accordingly.

Refs. TS-2516

Roffenlund committed:
Remove 404 error from readme API view

Remove the check which raises and error if package_version.readme is
None since readme cannot be null in the database.

Refs. TS-2516
